File: bratek20-spring/src/main/java/pl/bratek20/spring/app/SpringApp.java
package pl.bratek20.spring.app;

import com.github.bratek20.architecture.context.spring.SpringContext;
import lombok.RequiredArgsConstructor;
import org.springframework.boot.WebApplicationType;
import org.springframework.boot.builder.SpringApplicationBuilder;

@RequiredArgsConstructor
public class SpringApp {

    private final Class<?> configuration;
    private final String[] args;

    SpringContext run() {
        var configs = new Class<?>[]{SpringAppConfig.class};
        if (configuration != null) {
            configs = new Class<?>[]{configuration, SpringAppConfig.class};
        }
        var context = new SpringApplicationBuilder(configs)
            .web(WebApplicationType.NONE)
            .run(args);

        return new SpringContext(context);
    }

    public static SpringContext run(Class<?> config, String[] args) {
        var app = new SpringApp(config, args);
        return app.run();
    }
}

File: bratek20-spring/src/main/java/com/github/bratek20/spring/app/SpringApp.kt
package com.github.bratek20.spring.app

import com.github.bratek20.architecture.context.spring.SpringContext
import org.springframework.boot.WebApplicationType
import org.springframework.boot.builder.SpringApplicationBuilder

class SpringApp(
    private val configuration: Class<*>? = null,
    private val args: Array<String> = emptyArray()
) {

    fun run(): SpringContext {
        val configs = mutableListOf<Class<*>>(AppConfig::class.java)
        configuration?.let { configs.add(it) }

        val context = SpringApplicationBuilder(*configs.toTypedArray())
            .web(WebApplicationType.NONE)
            .run(*args)

        return SpringContext(context)
    }

    companion object {
        fun run(config: Class<*>? = null, args: Array<String> = emptyArray()): SpringContext {
            val app = SpringApp(config, args)
            return app.run()
        }
    }
}
